### PR TITLE
Update `kingpin` & allow autocompletion

### PR DESCRIPTION
**Pull request by gzdunek**
e counterpart https://github.com/gravitational/teleport.e/pull/1402

This PR bumps our kingpin to the updated fork version. History of added commits: https://github.com/gravitational/kingpin/pull/18/commits
and a small fix: https://github.com/gravitational/kingpin/pull/19/commits.

The only changes for the user that I noticed:
- bool flag now have `[no]` added e.g `--[no-]debug` (previously it was only `--debug`). I guess we can keep it?
-----------

The updated kingpin allows us to enable autocompletion. You can turn it on (for tsh or for any other binary) by adding the following lines to your:
- `.zshrc`:
```
autoload -Uz compinit
compinit

eval "$(tsh --completion-script-zsh)" 
``` 
(in case of a local build you should replace `tsh` with the actual path to it)

 - `.bashrc`
```
 eval "$(tsh --completion-script-bash)"
``` 
You can read more about the autocompletion here: https://github.com/gravitational/kingpin/blob/master/README.md#bashzsh-shell-completion. The README mentions another way of autocompletion script in **Installation by Package** section, but this requires more work and I think we can stay with the statements from above. I'm going to add them to our tsh docs in another PR.

Please note that we autocomplete only commands and flags. In the future we can add completing the actual values, like nodes or databases. 

The PR is quite large, because I had to change all the imports - kingpin is a Go module now.

### CODE REVIEW

File: e_imports.go [hunk 1]
```diff
@@ -43,6 +43,7 @@ xargs go list -find -f '{{if (and
 */
 
 import (
+	_ "github.com/alecthomas/kingpin/v2"
```
Comment (codingllama) [inline e_imports.go:46]: Note: this requires https://github.com/gravitational/teleport.e/pull/1402 to be true.

File: go.mod [hunk 2]
```diff
@@ -184,7 +183,10 @@ require (
 	github.com/mailgun/minheap v0.0.0-20170619185613-3dbe6c6bf55f // indirect
 )
 
-require cloud.google.com/go/compute/metadata v0.2.3
+require (
+	cloud.google.com/go/compute/metadata v0.2.3
```
Comment (codingllama) [inline go.mod:187]: Could you add those imports to the big "require" block above, the same that had kingpin before, and `go mod tidy` it again? It's easier if we keep all directs in a single block.

File: go.mod [hunk 3]
```diff
@@ -184,7 +183,10 @@ require (
...
+	github.com/alecthomas/kingpin/v2 v2.3.1
```
Comment (codingllama) [inline go.mod:188]: Please add this comment and make sure it survives any changes/tidying:

```suggestion
	github.com/alecthomas/kingpin/v2 v2.3.1 // replaced
```

This is useful to draw attention to bots trying to update forked dependencies.

File: lib/utils/cli.go [hunk 4]
```diff
@@ -466,60 +466,60 @@ const defaultCommandPrintfWidth = 12
...
-{{define "FormatCommands"}}\
-{{range .FlattenedCommands}}\
-{{if not .Hidden}}\
-  {{.FullCommand | printf "%%-%ds"}}{{if .Default}} (Default){{end}} {{ .Help }}
-{{end}}\
-{{end}}\
-{{end}}\
-
-{{define "FormatUsage"}}\
+const defaultUsageTemplate = `{{define "FormatCommand" -}}
```
Comment (codingllama) [inline lib/utils/cli.go:469]: Is this the only template affected? Should we chase all templates used in tsh?

**Review by codingllama — COMMENTED** [review]
LGTM, although I didn't test the binaries to compare behavior/output.

I suggest getting both this and https://github.com/gravitational/teleport.e/pull/1402 approved, landing the latter first and pulling the e/ update here, so the time window where we get go.mod errors (due to lack of alecthomas/kingpin/v2) is small.

Alternatively, we can play it safe and add both deps to go.mod, merge https://github.com/gravitational/teleport.e/pull/1402 and then merge this one with a cleanup. Less breakage potential, if I'm correct.

Ditto for ordering backports.



> bool flag now have [no] added e.g --[no-]debug (previously it was only --debug). I guess we can keep it?

IMO it's an improvement, I missed having a way to set negatives (like `--no-allow-passwordless`).

**Review by codingllama — APPROVED** [review]

File: .github/dependabot.yml [hunk 5]
```diff
@@ -20,6 +20,7 @@ updates:
       - dependency-name: github.com/go-mysql-org/go-mysql
       - dependency-name: github.com/go-redis/redis/v9
       - dependency-name: github.com/gogo/protobuf
+      - dependency-name: github.com/alecthomas/kingpin/v2
```
[the remark this answers is not in the document]
Comment (gzdunek) [reply .github/dependabot.yml]: Done

File: go.mod (hunk 2 above)
Comment (gzdunek) [reply go.mod]: Done

File: lib/utils/cli.go (hunk 4 above)
Comment (gzdunek) [reply lib/utils/cli.go:469]: Do we have any other templates? If I am correct, this is the only custom template we have.

**Review by codingllama — APPROVED** [review]
LGTM, but friendly reminder that we are heading to an e/ build breakage with this and https://github.com/gravitational/teleport.e/pull/1402, as for a period of time either the old or the new kingpin will be missing from go.mod.

For the sake of not breaking the build, should we push first a smaller PR that just adds "github.com/alecthomas/kingpin/v2" (and its replace) to go.mod, so we can keep the builds stable?

**Comment by gzdunek** [comment]
Sorry, I forgot to address this problem. Your idea of adding both dependencies sounds reasonable, I will push a PR for that.

**Comment by codingllama** [comment]
Awesome, thanks Grzegorz.

**Comment by codingllama** [comment]
@camscale, the replace is so the fork has only the minimal number of changes. A non-replaced import needs the fork's go.mod to say "gravitational/kingpin" instead of "alecthomas/kingpin", meaning it's less of a fork and more its "own" project.

It is typical for our forks to act as replaces, but I have no strong opinions it y'all want to do differently here.

**Review by zmb3 — APPROVED** [review]

**Review by camscale — APPROVED** [review]
Since the standard here seems to be to use replace, i'll withdraw my objection. Just wanted to express my view here.

**Pull request merged**